// use-view-transitions/dist/cjs/react-vt.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/react-vt.ts
var react_vt_exports = {};
__export(react_vt_exports, {
  AutoViewTransitionsOnClick: () => AutoViewTransitionsOnClick,
  SuspendViewTransition: () => SuspendViewTransition,
  useAutoViewTransitions: () => useAutoViewTransitions,
  useViewTransition: () => useViewTransition
});
module.exports = __toCommonJS(react_vt_exports);
var import_react = require("react");
var suspendersCount = 0;
var observers = /* @__PURE__ */ new Set();
var didCaptureNewState = null;
function suspendViewTransitionCapture() {
  suspendersCount++;
}
function resumeViewTransitionCapture() {
  !--suspendersCount;
  observers.forEach((observer) => observer());
}
var _a;
var areViewTransitionsSupported = typeof ((_a = globalThis.document) == null ? void 0 : _a.startViewTransition) === "function";
function useBlockRendering(blocked) {
  const [, forceRender] = (0, import_react.useReducer)((x) => x + 1, 0);
  if (blocked) {
    const deadline = performance.now() + 1;
    while (performance.now() < deadline) {
    }
  }
  (0, import_react.useEffect)(() => {
    if (blocked)
      forceRender();
  });
}
function AutoViewTransitionsOnClick({ match = "a[href]" }) {
  const [, startTransition] = (0, import_react.useTransition)();
  const { startViewTransition } = useViewTransition();
  (0, import_react.useEffect)(() => {
    if (!match || !globalThis.document)
      return;
    function captureClick(event) {
      const target = event.target;
      if (!target.matches(match) || !event.isTrusted)
        return;
      event.preventDefault();
      event.stopPropagation();
      startViewTransition(() => startTransition(() => target.click()));
    }
    globalThis.document.addEventListener("click", captureClick, { capture: true });
    return () => {
      globalThis.document.removeEventListener("click", captureClick);
    };
  }, []);
  return null;
}
function SuspendViewTransition() {
  (0, import_react.useEffect)(() => {
    suspendViewTransitionCapture();
    return () => resumeViewTransitionCapture();
  }, []);
  return null;
}
function useViewTransition() {
  const [transitionState, setTransitionState] = (0, import_react.useState)("idle");
  (0, import_react.useSyncExternalStore)((onStoreChange) => {
    observers.add(onStoreChange);
    return () => {
      observers.delete(onStoreChange);
    };
  }, () => suspendersCount, () => 0);
  (0, import_react.useEffect)(() => {
    if (didCaptureNewState && !suspendersCount) {
      didCaptureNewState();
      didCaptureNewState = null;
    }
  });
  useBlockRendering(transitionState === "capturing-old");
  function startViewTransition(updateCallback) {
    if (!areViewTransitionsSupported) {
      if (updateCallback)
        (0, import_react.startTransition)(updateCallback);
      return;
    }
    suspendViewTransitionCapture();
    setTransitionState("capturing-old");
    const transition = document.startViewTransition(() => new Promise((resolve) => __async(this, null, function* () {
      setTransitionState("capturing-new");
      resumeViewTransitionCapture();
      if (updateCallback)
        yield updateCallback();
      didCaptureNewState = resolve;
    })));
    transition.finished.then(() => {
      setTransitionState("idle");
    });
    transition.ready.then(() => {
      setTransitionState("animating");
    }).catch((e) => {
      console.error(e);
      setTransitionState("skipped");
    });
  }
  return {
    transitionState,
    startViewTransition,
    suspendViewTransitionCapture,
    resumeViewTransitionCapture
  };
}
function useAutoViewTransitions({ enabled = true }) {
  const { transitionState, startViewTransition } = useViewTransition();
  const [isPending] = (0, import_react.useTransition)();
  (0, import_react.useEffect)(() => {
    if (enabled && transitionState === "idle" && isPending && areViewTransitionsSupported) {
      startViewTransition();
    }
  }, [transitionState, isPending, enabled]);
}
